lightning/src/chain/keysinterface.rs

```diff
@@ -391,6 +392,12 @@ pub trait KeysInterface: Send + Sync {
 	/// contain no versioning scheme. You may wish to include your own version prefix and ensure
 	/// you've read all of the provided bytes to ensure no corruption occurred.
 	fn read_chan_signer(&self, reader: &[u8]) -> Result<Self::Signer, DecodeError>;
+
+	/// Sign an invoice's preimage (note that this is the preimage of the invoice, not the HTLC's
+	/// preimage). By parameterizing by the preimage instead of the hash, we allow implementors of
+	/// this trait to parse the invoice and make sure they're signing what they expect, rather than
+	/// blindly signing the hash.
+	fn sign_invoice(&self, invoice_preimage: Vec<u8>) -> Result<RecoverableSignature, DecodeError>;
```

Comment:
`DecodeError` definitely feels pretty strange here - I suppose in principle users may be decoding so that's a part of the possible errors, but users are also signing, and you could imagine something like "hard ware wallet unplugged" as an error. I guess we could do `()` for the error type  here like we do a few others - ultimately we're just gonna pass the user their own error back again.